// @jahia/moonstone/dist/components/Collapsible/UncontrolledCollapsible.js
import _extends from "@babel/runtime/helpers/extends";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["children", "onClick", "isDefaultExpanded"];
import React, { useState } from 'react';
import "./Collapsible.css";
import { ControlledCollapsible } from "./ControlledCollapsible";

var UncontrolledCollapsibleForwardRef = function UncontrolledCollapsibleForwardRef(_ref, ref) {
  var children = _ref.children,
      _ref$onClick = _ref.onClick,
      onClick = _ref$onClick === void 0 ? function () {
    return undefined;
  } : _ref$onClick,
      _ref$isDefaultExpande = _ref.isDefaultExpanded,
      isDefaultExpanded = _ref$isDefaultExpande === void 0 ? false : _ref$isDefaultExpande,
      other = _objectWithoutProperties(_ref, _excluded);

  var _useState = useState(isDefaultExpanded),
      _useState2 = _slicedToArray(_useState, 2),
      isExpanded = _useState2[0],
      setIsExpanded = _useState2[1];

  var handleOnClick = function handleOnClick(e) {
    setIsExpanded(!isExpanded);
    onClick(e);
  };

  return /*#__PURE__*/React.createElement(ControlledCollapsible, _extends({
    ref: ref,
    isExpanded: isExpanded,
    onClick: function onClick(e) {
      return handleOnClick(e);
    }
  }, other), children);
};

export var UncontrolledCollapsible = /*#__PURE__*/React.forwardRef(UncontrolledCollapsibleForwardRef);
UncontrolledCollapsible.displayName = 'UncontrolledCollapsible';
