// @jahia/moonstone/dist/layouts/content/LayoutContent.stories.js
import _extends from "@babel/runtime/helpers/extends";
import React from 'react';
import { LayoutContent } from "./..";
import { Header } from "../../components";
import { FakeContent } from "../../__storybook__/FakeComponents";
export default {
  title: 'Layouts/LayoutContent',
  component: LayoutContent,
  decorators: [function (StoryCmp) {
    return /*#__PURE__*/React.createElement("div", {
      style: {
        width: '1OOvw',
        height: '100vh',
        display: 'flex'
      }
    }, /*#__PURE__*/React.createElement(StoryCmp, null));
  }],
  argTypes: {
    header: {
      control: false
    },
    content: {
      control: false
    }
  }
};

var Template = function Template(args) {
  return /*#__PURE__*/React.createElement(LayoutContent, _extends({
    header: /*#__PURE__*/React.createElement(Header, {
      title: "Header"
    }),
    content: /*#__PURE__*/React.createElement(FakeContent, null)
  }, args));
};

export var Default = Template.bind({});
export var Centered = Template.bind({});
Centered.args = {
  isCentered: true
};
export var WithoutPadding = Template.bind({});
WithoutPadding.args = {
  hasPadding: false
};
export var Loading = Template.bind({});
Loading.args = {
  isLoading: true
};
