// @jahia/moonstone/dist/components/RadioGroup/RadioItem/RadioItem.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["className", "id", "value", "label", "description", "isDisabled", "isReadOnly"];
import React from 'react';
import clsx from 'clsx';
import "./RadioItem.css";
import { RadioChecked, RadioUnchecked } from "../../../icons";
import { Typography } from "../..";
import { RadioGroupContext } from "../RadioGroup.context";
export var RadioItem = function RadioItem(_ref) {
  var className = _ref.className,
      id = _ref.id,
      value = _ref.value,
      label = _ref.label,
      description = _ref.description,
      isDisabled = _ref.isDisabled,
      isReadOnly = _ref.isReadOnly,
      props = _objectWithoutProperties(_ref, _excluded);

  var context = React.useContext(RadioGroupContext);
  var isDisabledItem = typeof context.isDisabled === 'undefined' ? isDisabled : context.isDisabled;
  var isReadOnlyItem = typeof context.isReadOnly === 'undefined' ? isReadOnly : context.isReadOnly;
  return /*#__PURE__*/React.createElement(Typography, {
    className: clsx('moonstone-radio-container flexCol', className),
    "aria-readonly": isReadOnlyItem,
    "aria-disabled": isDisabledItem,
    variant: "body",
    weight: "default",
    component: "label"
  }, /*#__PURE__*/React.createElement("div", {
    className: clsx('flexRow alignCenter')
  }, /*#__PURE__*/React.createElement("div", {
    className: clsx('moonstone-radio')
  }, /*#__PURE__*/React.createElement("input", _extends({}, props, {
    className: clsx('moonstone-radio_input'),
    type: "radio",
    checked: context.value === value,
    disabled: isDisabledItem,
    name: context.name,
    "aria-readonly": isReadOnlyItem,
    id: id,
    value: value,
    "aria-labelledby": "".concat(id, "-label"),
    "aria-describedby": description ? "".concat(id, "-description") : null,
    onChange: function onChange(event) {
      context.onChange(event, value);
    }
  })), /*#__PURE__*/React.createElement(RadioChecked, {
    className: clsx('moonstone-radio_icon moonstone-radio_iconChecked')
  }), /*#__PURE__*/React.createElement(RadioUnchecked, {
    className: clsx('moonstone-radio_icon moonstone-radio_iconUnchecked')
  })), /*#__PURE__*/React.createElement(Typography, {
    id: "".concat(id, "-label"),
    variant: "body",
    weight: "default",
    component: "span",
    className: clsx('moonstone-radio-label')
  }, label)), description && /*#__PURE__*/React.createElement(Typography, {
    id: "".concat(id, "-description"),
    variant: "caption",
    weight: "default",
    component: "span",
    className: clsx('moonstone-radio-description flexRow')
  }, description));
};
RadioItem.displayName = 'RadioItem';
