// @jahia/moonstone/dist/components/TreeView/TreeView.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["openedItems", "defaultOpenedItems", "onOpenItem", "onCloseItem", "data"];
import React from 'react';
import { UncontrolledTreeView } from "./UncontrolledTreeView";
import { ControlledTreeView } from "./ControlledTreeView";

var TreeViewForwardRef = function TreeViewForwardRef(_ref, ref) {
  var openedItems = _ref.openedItems,
      defaultOpenedItems = _ref.defaultOpenedItems,
      onOpenItem = _ref.onOpenItem,
      onCloseItem = _ref.onCloseItem,
      data = _ref.data,
      others = _objectWithoutProperties(_ref, _excluded);

  // If no data render nothing
  if (!Array.isArray(data) || data.length < 1) {
    return null;
  }

  if (typeof openedItems === 'undefined') {
    return /*#__PURE__*/React.createElement(UncontrolledTreeView, _extends({
      defaultOpenedItems: defaultOpenedItems,
      data: data
    }, others));
  }

  return /*#__PURE__*/React.createElement(ControlledTreeView, _extends({
    ref: ref,
    openedItems: openedItems,
    data: data,
    onOpenItem: onOpenItem,
    onCloseItem: onCloseItem
  }, others));
};

export var TreeView = /*#__PURE__*/React.forwardRef(TreeViewForwardRef);
TreeView.displayName = 'TreeView';
