// @jahia/moonstone/dist/components/Separator/Separator.types.js
export var SeparatorSpacings;

(function (SeparatorSpacings) {
  SeparatorSpacings["None"] = "none";
  SeparatorSpacings["Small"] = "small";
  SeparatorSpacings["Medium"] = "medium";
  SeparatorSpacings["Big"] = "big";
})(SeparatorSpacings || (SeparatorSpacings = {}));

export var SeparatorSizes;

(function (SeparatorSizes) {
  SeparatorSizes["Medium"] = "medium";
  SeparatorSizes["Large"] = "large";
  SeparatorSizes["Full"] = "full";
})(SeparatorSizes || (SeparatorSizes = {}));

export var SeparatorVariants;

(function (SeparatorVariants) {
  SeparatorVariants["Horizontal"] = "horizontal";
  SeparatorVariants["Vertical"] = "vertical";
})(SeparatorVariants || (SeparatorVariants = {}));

export var SeparatorInvisible;

(function (SeparatorInvisible) {
  SeparatorInvisible["FirstChild"] = "firstChild";
  SeparatorInvisible["LastChild"] = "lastChild";
  SeparatorInvisible["OnlyChild"] = "onlyChild";
  SeparatorInvisible["FirstOrLastChild"] = "firstOrLastChild";
})(SeparatorInvisible || (SeparatorInvisible = {}));
