// @jahia/ui-extender/dist/js/actions/samples/ButtonRenderer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ButtonRenderer = void 0;
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _excluded = ["onClick", "isVisible", "enabled", "label"];
var ButtonRenderer = function ButtonRenderer(_ref) {
  var _onClick = _ref.onClick,
    isVisible = _ref.isVisible,
    enabled = _ref.enabled,
    label = _ref.label,
    props = (0, _objectWithoutProperties2["default"])(_ref, _excluded);
  return isVisible !== false && /*#__PURE__*/_react["default"].createElement("button", {
    style: {
      color: '#444444',
      background: '#F3F3F3',
      padding: '5px 10px',
      borderRadius: '2px',
      fontWeight: 'bold',
      fontSize: '9pt',
      outline: 'none'
    },
    disabled: enabled === false,
    type: "button",
    onClick: function onClick(ev) {
      return _onClick(props, ev);
    }
  }, label);
};
exports.ButtonRenderer = ButtonRenderer;
ButtonRenderer.propTypes = {
  onClick: _propTypes["default"].func,
  isVisible: _propTypes["default"].bool,
  // eslint-disable-next-line react/boolean-prop-naming
  enabled: _propTypes["default"].bool,
  label: _propTypes["default"].string
};
