// @jahia/ui-extender/dist/js/actions/samples/Modal.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Modal = void 0;
var _propTypes = _interopRequireDefault(require("prop-types"));
var _react = _interopRequireDefault(require("react"));
var Modal = function Modal(_ref) {
  var context = _ref.context,
    onExited = _ref.onExited;
  return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement("div", {
    className: "backdrop",
    style: {
      position: 'absolute',
      width: '100%',
      height: '100%',
      top: 0,
      left: 0,
      opacity: 0.1,
      backgroundColor: 'black'
    },
    onClick: onExited
  }), /*#__PURE__*/_react["default"].createElement("div", {
    className: "modal",
    style: {
      position: 'absolute',
      width: '200px',
      height: '40px',
      top: 'calc( 50% - 20px )',
      left: 'calc( 50% - 100px )',
      border: '1px solid',
      backgroundColor: 'white',
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center'
    }
  }, /*#__PURE__*/_react["default"].createElement("div", {
    style: {
      flex: '0 1 auto'
    }
  }, context.content)));
};
exports.Modal = Modal;
Modal.propTypes = {
  context: _propTypes["default"].object.isRequired,
  onExited: _propTypes["default"].func.isRequired
};
