// @jahia/data-helper/dist/esm/hooks/useSiteInfo/useSiteInfo.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
import { useQuery } from '@apollo/client';
import { SITE_INFO_QUERY } from "./useSiteInfo.gql-queries";
var adaptSiteInfo = function adaptSiteInfo(data) {
  if (data && (data.jcr || data.wsDefault)) {
    var res = data.jcr ? data.jcr.result : data.wsDefault.result;
    if (res) {
      return _objectSpread({}, res.site);
    }
  }
  return {
    languages: []
  };
};
export var useSiteInfo = function useSiteInfo(_ref) {
  var siteKey = _ref.siteKey,
    displayLanguage = _ref.displayLanguage;
  var variables = {
    path: '/sites/' + siteKey,
    displayLanguage: displayLanguage
  };
  var res = useQuery(SITE_INFO_QUERY, {
    variables: variables,
    errorPolicy: 'ignore'
  });
  return _objectSpread(_objectSpread({}, res), {}, {
    siteInfo: adaptSiteInfo(res.data)
  });
};
