// @jahia/moonstone/dist/components/RadioGroup/ControlledRadioGroup.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["children", "name", "value", "isDisabled", "isReadOnly", "onChange", "className"];
import React from 'react';
import clsx from 'clsx';
import { RadioGroupContext } from "./RadioGroup.context";
export var ControlledRadioGroup = function ControlledRadioGroup(_ref) {
  var children = _ref.children,
      name = _ref.name,
      value = _ref.value,
      isDisabled = _ref.isDisabled,
      isReadOnly = _ref.isReadOnly,
      onChange = _ref.onChange,
      className = _ref.className,
      props = _objectWithoutProperties(_ref, _excluded);

  if (typeof value === 'undefined' || value === '') {
    value = children[0].props.value;
  }

  var provider = {
    name: name,
    value: value,
    isDisabled: isDisabled,
    isReadOnly: isReadOnly,
    onChange: onChange
  };
  return /*#__PURE__*/React.createElement(RadioGroupContext.Provider, {
    value: provider
  }, /*#__PURE__*/React.createElement("div", _extends({}, props, {
    className: clsx('flexCol', className)
  }), children));
};
