// @jahia/ui-extender/dist/esm/actions/core/DisplayAction.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _pt from "prop-types";
var _excluded = ["context"],
  _excluded2 = ["context", "actionKey", "render", "loading"];
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
import React from 'react';
import { registry } from "../../registry";
var count = 0;
var getRenderWrapper = function getRenderWrapper(Render) {
  var RenderWrapper = function RenderWrapper(_ref) {
    var context = _ref.context,
      otherProps = _objectWithoutProperties(_ref, _excluded);
    return /*#__PURE__*/React.createElement(Render, _extends({}, context, otherProps));
  };
  return RenderWrapper;
};
var DisplayAction = /*#__PURE__*/function (_React$PureComponent) {
  _inherits(DisplayAction, _React$PureComponent);
  var _super = _createSuper(DisplayAction);
  function DisplayAction(props) {
    var _this;
    _classCallCheck(this, DisplayAction);
    _this = _super.call(this, props);
    _this.id = props.actionKey + '-' + count++;
    _this.RenderWrapper = getRenderWrapper(props.render);
    return _this;
  }
  _createClass(DisplayAction, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
        context = _this$props.context,
        actionKey = _this$props.actionKey,
        render = _this$props.render,
        loading = _this$props.loading,
        otherProps = _objectWithoutProperties(_this$props, _excluded2);
      var action = registry.get('action', actionKey);
      if (!action) {
        return null;
      }
      if (context) {
        console.warn('Warn : context in DisplayAction is deprecated', actionKey, context);
      }
      var Component = typeof action.component === 'function' ? action.component : render;

      // Merge props and context. To remove when context is not supported anymore
      var mergedProps = _objectSpread(_objectSpread({}, context), otherProps);
      var componentProps = _objectSpread(_objectSpread(_objectSpread({}, action), mergedProps), {}, {
        originalContext: mergedProps,
        id: this.id,
        actionKey: actionKey
      });
      if (componentProps.init) {
        componentProps.init(componentProps, this.props);
      }
      var info = {};
      info['data-registry-key'] = action.type + ':' + action.key;
      if (otherProps.target && action.targets) {
        var foundTarget = action.targets.find(function (t) {
          return t.id === otherProps.target;
        });
        if (foundTarget) {
          info['data-registry-target'] = foundTarget.id + ':' + foundTarget.priority;
        }
      }

      // Props are passed as as context and props. Context can be removed when not supported anymore
      return /*#__PURE__*/React.createElement(Component, _extends({
        key: this.id
      }, componentProps, {
        buttonProps: _objectSpread(_objectSpread({}, componentProps.buttonProps), info),
        context: componentProps,
        render: this.RenderWrapper,
        loading: loading
      }));
    }
  }]);
  return DisplayAction;
}(React.PureComponent);
_defineProperty(DisplayAction, "propTypes", {
  /**
       * The key of the action to display
       */
  actionKey: _pt.string.isRequired,
  /**
       * The action context
       */
  context: _pt.object
});
export { DisplayAction };
