// @jahia/ui-extender/dist/js/adminRoutes/useAdminRoutesTreeStructure.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useAdminRouteTreeStructure = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _react = require("react");
var _registry = require("../registry");
var _IframeRenderer = require("../IframeRenderer");
var _Tree = require("./Tree");
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var useAdminRouteTreeStructure = function useAdminRouteTreeStructure(target, selected) {
  var result = (0, _react.useMemo)(function () {
    var getAllRoutes = function getAllRoutes(baseTarget) {
      var parent = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';
      return _registry.registry.find({
        type: 'adminRoute',
        target: baseTarget + parent
      }).flatMap(function (route) {
        return [route].concat((0, _toConsumableArray2["default"])(getAllRoutes(baseTarget, '-' + route.key)));
      }).map(function (route) {
        return _objectSpread(_objectSpread({}, route), {}, {
          render: route.render || typeof route.iframeUrl === 'string' && function () {
            return (0, _IframeRenderer.getIframeRenderer)(route.iframeUrl);
          }
        });
      });
    };
    var routes = getAllRoutes(target);
    var createTree = function createTree(baseTarget) {
      var parent = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';
      return _registry.registry.find({
        type: 'adminRoute',
        target: baseTarget + parent
      }).filter(function (route) {
        return !route.omitFromTree;
      }).map(function (route) {
        return _objectSpread(_objectSpread({}, route), {}, {
          children: createTree(baseTarget, '-' + route.key)
        });
      });
    };
    var tree = new _Tree.Tree(createTree(target));
    var allPermissions = routes.filter(function (route) {
      return route.requiredPermission;
    }).map(function (route) {
      return route.requiredPermission;
    }).filter(function (item, pos, self) {
      return self.indexOf(item) === pos;
    });
    return {
      routes: routes,
      tree: tree,
      allPermissions: allPermissions
    };
  }, [target]);
  var tree = result.tree,
    routes = result.routes,
    allPermissions = result.allPermissions;
  var defaultOpenedItems = [];
  if (selected) {
    var selectedItem = _registry.registry.get('adminRoute', selected);
    while (selectedItem) {
      var parentTarget = selectedItem.targets.find(function (t) {
        return t.id.startsWith(target + '-');
      });
      if (parentTarget) {
        var parent = parentTarget.id.substr(target.length + 1);
        defaultOpenedItems.push(parent);
        selectedItem = _registry.registry.get('adminRoute', parent);
      } else {
        selectedItem = null;
      }
    }
  }
  return {
    tree: tree,
    defaultOpenedItems: defaultOpenedItems,
    routes: routes,
    allPermissions: allPermissions
  };
};
exports.useAdminRouteTreeStructure = useAdminRouteTreeStructure;
