// @jahia/moonstone/dist/components/Dropdown/TreeViewMenu.js
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import React, { useState } from 'react';
import { usePositioning } from "../../hooks/usePositioning";
import { SearchInput, TreeView } from "./..";
import "../Menu/Menu.css";
import clsx from 'clsx';

function filterNodes(predicate, nodes, opened) {
  var filtered = [];
  nodes.forEach(function (c) {
    var filterResult = filterNode(predicate, c, opened);

    if (filterResult) {
      filtered.push(filterResult);
    }
  });
  return filtered;
}

var filterNode = function filterNode(predicate, node, opened) {
  var match = predicate(node);
  var children = [];

  if (node.children) {
    var filteredChildren = filterNodes(predicate, node.children, opened);

    if (filteredChildren.length > 0) {
      children.push.apply(children, _toConsumableArray(filteredChildren));
      opened.push(node.id);
    }
  }

  if (match || children.length > 0) {
    return _objectSpread(_objectSpread({}, node), {}, {
      treeItemProps: {
        className: clsx({
          'moonstone-disabled': !match
        })
      },
      isDisabled: !match,
      children: children
    });
  }
};

var find = function find(predicate, data, opened) {
  if (predicate(data)) {
    return data;
  }

  if (data.children) {
    var res = data.children.reduce(function (current, child) {
      return current || find(predicate, child, opened);
    }, null);

    if (res && opened) {
      opened.push(data.id);
    }

    return res;
  }
};

var flatten = function flatten(data) {
  var _data$forEach;

  var res = [];

  var fn = function fn(current) {
    res.push(current);

    if (current.children) {
      current.children.forEach(fn);
    }
  };

  data === null || data === void 0 ? void 0 : (_data$forEach = data.forEach) === null || _data$forEach === void 0 ? void 0 : _data$forEach.call(data, fn);
  return res;
};

export var TreeViewMenu = function TreeViewMenu(_ref) {
  var isDisplayed = _ref.isDisplayed,
      minWidth = _ref.minWidth,
      maxWidth = _ref.maxWidth,
      maxHeight = _ref.maxHeight,
      anchorEl = _ref.anchorEl,
      anchorPosition = _ref.anchorPosition,
      anchorElOrigin = _ref.anchorElOrigin,
      transformElOrigin = _ref.transformElOrigin,
      position = _ref.position,
      hasOverlay = _ref.hasOverlay,
      hasSearch = _ref.hasSearch,
      autoAddSearchLimit = _ref.autoAddSearchLimit,
      treeData = _ref.treeData,
      value = _ref.value,
      values = _ref.values,
      handleSelect = _ref.handleSelect,
      onClose = _ref.onClose;

  var _usePositioning = usePositioning(isDisplayed, anchorPosition, anchorEl, anchorElOrigin, transformElOrigin, position),
      _usePositioning2 = _slicedToArray(_usePositioning, 2),
      stylePosition = _usePositioning2[0],
      itemRef = _usePositioning2[1]; // UseEnterExitCallbacks(isDisplayed, onExiting, onExited, onEntering, onEntered);


  var _useState = useState(''),
      _useState2 = _slicedToArray(_useState, 2),
      inputValue = _useState2[0],
      setInputValue = _useState2[1];

  var _useState3 = useState([]),
      _useState4 = _slicedToArray(_useState3, 2),
      openedItems = _useState4[0],
      setOpenedItems = _useState4[1];

  var onOpenItem = function onOpenItem(node) {
    setOpenedItems(function (previousOpenedItems) {
      return [].concat(_toConsumableArray(previousOpenedItems), [node.id]);
    });
  };

  var onCloseItem = function onCloseItem(node) {
    setOpenedItems(function (previousOpenedItems) {
      return previousOpenedItems.filter(function (item) {
        return item !== node.id;
      });
    });
  };

  var openedBySearch = [];
  var selected = [];

  if (inputValue !== '') {
    treeData = filterNodes(function (node) {
      return node.label.toLowerCase().includes(inputValue.toLowerCase());
    }, treeData, openedBySearch);
  }

  if (value) {
    treeData.forEach(function (single) {
      var item = find(function (data) {
        return data.value === value;
      }, single, openedBySearch);

      if (item) {
        selected.push(item.id);
      }
    });
  }

  if (values) {
    values.forEach(function (v) {
      treeData.forEach(function (single) {
        var item = find(function (data) {
          return data.value === v;
        }, single, openedBySearch);

        if (item) {
          selected.push(item.id);
        }
      });
    });
  } // ---
  // Styling
  // ---


  var styleMenu = _objectSpread({
    position: position
  }, stylePosition);

  if (minWidth) {
    styleMenu.minWidth = minWidth;
  }

  if (maxWidth) {
    styleMenu.maxWidth = maxWidth;
  }

  if (maxHeight) {
    styleMenu.maxHeight = maxHeight;
  }

  var hasAutoSearch = hasSearch;

  if (typeof hasSearch === 'undefined') {
    var _flatten;

    hasAutoSearch = ((_flatten = flatten(treeData)) === null || _flatten === void 0 ? void 0 : _flatten.length) > autoAddSearchLimit;
  } // ---
  // Render
  // ---


  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("menu", {
    ref: itemRef,
    className: "moonstone-menu",
    style: styleMenu
  }, hasAutoSearch && /*#__PURE__*/React.createElement("div", {
    className: "moonstone-menu_searchInput"
  }, /*#__PURE__*/React.createElement(SearchInput, {
    value: inputValue,
    onChange: function onChange(e) {
      return setInputValue(e.target.value);
    },
    onKeyPress: function onKeyPress(e) {
      if (e.key === 'Enter' && treeData.length > 0) {
        var item = find(function (data) {
          return !data.isDisabled;
        }, treeData[0]);

        if (item) {
          handleSelect(e, item);
        }
      }
    },
    onClear: function onClear() {
      return setInputValue('');
    }
  })), /*#__PURE__*/React.createElement(TreeView, {
    data: treeData,
    selectedItems: selected,
    size: "small",
    showCheckbox: Boolean(values),
    openedItems: [].concat(_toConsumableArray(openedItems), openedBySearch),
    onOpenItem: onOpenItem,
    onCloseItem: onCloseItem,
    onClickItem: function onClickItem(node, e) {
      handleSelect(e, node);
    }
  })), hasOverlay && isDisplayed && /*#__PURE__*/React.createElement("div", {
    "aria-hidden": "true",
    className: "moonstone-menu_overlay",
    onClick: onClose,
    onContextMenu: onClose
  }));
}; // Kept defaultProps here because of unnecessary re-rendering when provided as default parameters to the function component

/* eslint-disable react/default-props-match-prop-types */

TreeViewMenu.defaultProps = {
  hasOverlay: true,
  autoAddSearchLimit: 7,
  searchEmptyText: 'No results found.',
  position: 'fixed',
  anchorEl: null,
  anchorPosition: {
    top: 0,
    left: 0
  },
  anchorElOrigin: {
    vertical: 'bottom',
    horizontal: 'left'
  },
  transformElOrigin: {
    vertical: 'top',
    horizontal: 'left'
  }
};
/* eslint-enable react/default-props-match-prop-types */

TreeViewMenu.displayName = 'TreeViewMenu';
