// @jahia/data-helper/dist/js/hooks/useContentPreview/useContentPreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useContentPreview = void 0;
var _client = require("@apollo/client");
var _useContentPreview = require("./useContentPreview.gql-queries");
var useContentPreview = function useContentPreview(_ref) {
  var path = _ref.path,
    workspace = _ref.workspace,
    language = _ref.language,
    templateType = _ref.templateType,
    view = _ref.view,
    contextConfiguration = _ref.contextConfiguration,
    requestAttributes = _ref.requestAttributes,
    fetchPolicy = _ref.fetchPolicy;
  var variables = {
    path: path,
    templateType: templateType,
    view: view,
    contextConfiguration: contextConfiguration,
    language: language,
    workspace: workspace.toUpperCase(),
    requestAttributes: requestAttributes
  };
  return (0, _client.useQuery)(_useContentPreview.CONTENT_PREVIEW_QUERY, {
    variables: variables,
    errorPolicy: 'all',
    fetchPolicy: fetchPolicy
  });
};
exports.useContentPreview = useContentPreview;
