// @jahia/moonstone/dist/components/SecondaryNav/SecondaryNav.js
import _extends from "@babel/runtime/helpers/extends";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["header", "children", "isDefaultVisible", "isReversed", "onToggled", "className"];
import React, { useState } from 'react';
import clsx from 'clsx';
import "./SecondaryNav.css";
import { ResizableBox } from "../ResizableBox";
import { ChevronDoubleRight, ChevronDoubleLeft } from "../../icons";
export var SecondaryNav = function SecondaryNav(_ref) {
  var header = _ref.header,
      children = _ref.children,
      _ref$isDefaultVisible = _ref.isDefaultVisible,
      isDefaultVisible = _ref$isDefaultVisible === void 0 ? true : _ref$isDefaultVisible,
      _ref$isReversed = _ref.isReversed,
      isReversed = _ref$isReversed === void 0 ? true : _ref$isReversed,
      _ref$onToggled = _ref.onToggled,
      onToggled = _ref$onToggled === void 0 ? function () {
    return undefined;
  } : _ref$onToggled,
      className = _ref.className,
      props = _objectWithoutProperties(_ref, _excluded);

  var _useState = useState(isDefaultVisible),
      _useState2 = _slicedToArray(_useState, 2),
      isVisible = _useState2[0],
      setIsVisible = _useState2[1];

  var handleToggle = function handleToggle(e) {
    setIsVisible(function (prevState) {
      return !prevState;
    });
    onToggled(e);
  };

  return /*#__PURE__*/React.createElement(ResizableBox, _extends({
    className: clsx(className, 'flexFluid', 'flexCol_nowrap', 'moonstone-secondaryNav', {
      'moonstone-reversed': isReversed,
      'moonstone-secondaryNav_hidden': !isVisible
    }),
    enable: ['right'],
    size: isVisible ? null : {
      height: '0%',
      width: 0
    },
    minWidth: isVisible ? 245 : 0,
    maxWidth: "900",
    defaultSize: {
      height: '0%',
      width: '245px'
    }
  }, props), /*#__PURE__*/React.createElement("button", {
    type: "button",
    className: clsx('moonstone-secondaryNav_buttonToggle', {
      'moonstone-secondaryNav_buttonToggle_reversed': isReversed
    }),
    onClick: handleToggle
  }, isVisible && /*#__PURE__*/React.createElement(ChevronDoubleLeft, null), !isVisible && /*#__PURE__*/React.createElement(ChevronDoubleRight, null)), /*#__PURE__*/React.createElement("div", {
    className: clsx('moonstone-secondaryNav_wrapper', 'flexFluid', 'flexCol_nowrap')
  }, header, /*#__PURE__*/React.createElement("div", {
    className: clsx('flexFluid', 'flexCol_nowrap')
  }, children)));
};
SecondaryNav.displayName = 'SecondaryNav';
