// @jahia/data-helper/dist/js/legacy/Picker.gql-fragments.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PickerItemsFragment = void 0;
var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));
var _graphqlTag = _interopRequireDefault(require("graphql-tag"));
var _templateObject, _templateObject2, _templateObject3;
var PickerItemsFragment = {
  mixinTypes: {
    applyFor: 'node',
    gql: (0, _graphqlTag["default"])(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["fragment MixinTypes on JCRNode {\n            mixinTypes {\n                name\n            }\n        }"])))
  },
  isPublished: {
    applyFor: 'node',
    variables: {
      language: 'String!'
    },
    gql: (0, _graphqlTag["default"])(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["fragment PublicationStatus on JCRNode {\n            publicationStatus: aggregatedPublicationInfo(language: $language) {\n                publicationStatus\n            }\n        }"])))
  },
  primaryNodeType: {
    applyFor: 'node',
    gql: (0, _graphqlTag["default"])(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2["default"])(["fragment PrimaryNodeTypeName on JCRNode {\n            primaryNodeType {\n                name\n            }\n        }"])))
  }
};
exports.PickerItemsFragment = PickerItemsFragment;
