// @jahia/data-helper/dist/js/legacy/Picker.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Picker = void 0;
var _toArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toArray"));
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _react = _interopRequireDefault(require("react"));
var _reactComponents = require("@apollo/react-components");
var _graphqlTag = _interopRequireDefault(require("graphql-tag"));
var _fragments = require("../fragments");
var _Picker = require("./Picker.gql-fragments");
var _rfdc = _interopRequireDefault(require("rfdc"));
var _fastDeepEqual = _interopRequireDefault(require("fast-deep-equal"));
var _templateObject;
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } } // eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck
var clone = (0, _rfdc["default"])();
var Picker = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2["default"])(Picker, _React$Component);
  var _super = _createSuper(Picker);
  function Picker(props) {
    var _this;
    (0, _classCallCheck2["default"])(this, Picker);
    _this = _super.call(this, props);
    var fragments = props.fragments,
      rootPaths = props.rootPaths,
      onOpenItem = props.onOpenItem,
      onSelectItem = props.onSelectItem,
      openPaths = props.openPaths,
      selectedPaths = props.selectedPaths,
      defaultSelectedPaths = props.defaultSelectedPaths,
      onSelectionChange = props.onSelectionChange,
      defaultOpenPaths = props.defaultOpenPaths;
    var resolvedFragments = fragments || [_Picker.PickerItemsFragment.mixinTypes, _Picker.PickerItemsFragment.primaryNodeType, _Picker.PickerItemsFragment.isPublished, _fragments.displayName];
    _this.query = (0, _graphqlTag["default"])(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n            query PickerQuery($rootPaths:[String!]!, $selectable:[String]!, $openable:[String]!, $openPaths:[String!]!, $types:[String]!) {\n                jcr {\n                    rootNodes:nodesByPath(paths: $rootPaths) {\n                        name\n                        children(typesFilter:{types:$types}, limit:1) {\n                            pageInfo {\n                                nodesCount\n                            }\n                        }\n                        selectable : isNodeType(type: {types:$selectable})\n                        openable : isNodeType(type: {types:$openable})\n                        ... NodeCacheRequiredFields\n                        ... node\n                    },\n                    openNodes:nodesByPath(paths: $openPaths) {\n                        ... NodeCacheRequiredFields\n                        children(typesFilter:{types:$types}) {\n                            nodes {\n                                name\n                                children(typesFilter:{types:$types}, limit:1) {\n                                    pageInfo {\n                                        nodesCount\n                                    }\n                                }\n                                selectable : isNodeType(type: {types:$selectable})\n                                openable : isNodeType(type: {types:$openable})\n                                ... NodeCacheRequiredFields\n                                ... node\n                            }\n                        }\n                    }\n                }\n            }\n        ", ""])), _fragments.nodeCacheRequiredFields.gql);
    _this.query = (0, _fragments.replaceFragmentsInDocument)(_this.query, resolvedFragments);
    var state = {};
    _this.eventsHandlers = {};
    if (!openPaths) {
      // Uncontrolled mode
      state.isOpenControlled = false;
      state.openPaths = [];
      _this.eventsHandlers.onOpenItem = function (path, open) {
        _this.setState(function (prevState) {
          return {
            openPaths: open ? [].concat((0, _toConsumableArray2["default"])(prevState.openPaths), [path]) : prevState.openPaths.filter(function (thispath) {
              return thispath !== path;
            })
          };
        });
      };
      if (defaultOpenPaths) {
        state.openPaths = _this.addPathToOpenPath(defaultOpenPaths, rootPaths, state.openPaths);
      }
    } else {
      state.isOpenControlled = true;
      if (onOpenItem) {
        _this.eventsHandlers.onOpenItem = onOpenItem;
      }
    }
    if (!selectedPaths) {
      // Uncontrolled mode
      state.isSelectControlled = false;
      state.selectedPaths = defaultSelectedPaths ? clone(defaultSelectedPaths) : [];
      // Open selected path if open is uncontrolled
      if (defaultSelectedPaths && !state.isOpenControlled) {
        state.openPaths = _this.addPathToOpenPath(defaultSelectedPaths, rootPaths, state.openPaths);
      }
      _this.eventsHandlers.onSelectItem = function (path, selected, multiple) {
        _this.setState(function (prevState) {
          var newSelectedPaths = selected ? [].concat((0, _toConsumableArray2["default"])(multiple ? prevState.selectedPaths : []), [path]) : prevState.selectedPaths.filter(function (thispath) {
            return thispath !== path;
          });
          onSelectionChange(newSelectedPaths);
          return {
            selectedPaths: newSelectedPaths
          };
        });
      };
    } else if (onSelectItem) {
      state.isSelectControlled = true;
      if (onSelectItem) {
        _this.eventsHandlers.onSelectItem = onSelectItem;
      }
    }
    _this.state = state;

    // Binding
    _this.openPaths = _this.openPaths.bind((0, _assertThisInitialized2["default"])(_this));
    return _this;
  }
  (0, _createClass2["default"])(Picker, [{
    key: "getVariables",
    value: function getVariables(selectedPaths, openPaths) {
      var _this$props = this.props,
        rootPaths = _this$props.rootPaths,
        openableTypes = _this$props.openableTypes,
        selectableTypes = _this$props.selectableTypes,
        queryVariables = _this$props.queryVariables;
      var vars = {
        rootPaths: rootPaths,
        types: (0, _toConsumableArray2["default"])(new Set([].concat((0, _toConsumableArray2["default"])(openableTypes), (0, _toConsumableArray2["default"])(selectableTypes)))),
        selectable: selectableTypes,
        openable: openableTypes,
        openPaths: openPaths
      };
      if (queryVariables) {
        Object.assign(vars, queryVariables);
      }
      return vars;
    }
  }, {
    key: "getPickerEntries",
    value: function getPickerEntries(data, selectedPaths, openPaths) {
      var _this2 = this;
      var pickerEntries = [];
      var nodesById = {};
      var jcr = data.jcr;
      var addNode = function addNode(node, depth, index) {
        var selected = false;
        if (node.selectable) {
          selected = selectedPaths.indexOf(node.path) !== -1;
        }
        var pickerNode = {
          name: node.name,
          path: node.path,
          open: node.openable && openPaths.indexOf(node.path) !== -1,
          selected: selected,
          openable: node.openable,
          selectable: node.selectable,
          depth: depth,
          prefix: '&nbsp;'.repeat(depth * 3),
          node: node,
          hidden: false,
          hasChildren: node.children.pageInfo.nodesCount > 0
        };
        pickerEntries.splice(index, 0, pickerNode);
        nodesById[node.uuid] = pickerNode;
        return pickerNode;
      };
      if (jcr) {
        if (jcr.rootNodes) {
          jcr.rootNodes.forEach(function (rootNode) {
            var root = addNode(rootNode, 0, 0);
            root.hidden = _this2.props.hideRoot || false;
          });
        }
        if (jcr.openNodes) {
          jcr.openNodes.concat().sort(function (a, b) {
            return a.path > b.path ? 1 : b.path > a.path ? -1 : 0;
          }).forEach(function (node) {
            var parent = nodesById[node.uuid];
            if (parent) {
              var parentIndex = pickerEntries.indexOf(parent);
              node.children.nodes.slice().reverse().forEach(function (child) {
                addNode(child, parent.depth + 1, parentIndex + 1);
              });
            }
          });
        }
      }

      // Nodes loaded, fill selection list
      var selectedNodes = pickerEntries.filter(function (node) {
        return node.selected;
      }).map(function (node) {
        return node.node;
      });
      selectedPaths = selectedNodes.map(function (s) {
        return s.path;
      });
      pickerEntries = pickerEntries.filter(function (pickerNode) {
        return !pickerNode.hidden;
      });
      return pickerEntries;
    }
  }, {
    key: "addPathToOpenPath",
    value: function addPathToOpenPath(pathsToOpen, rootPaths, openPaths) {
      pathsToOpen.forEach(function (path) {
        var rootFound = false;
        if (!path.endsWith('/')) {
          path += '/';
        }
        var _path$split = path.split('/'),
          _path$split2 = (0, _toArray2["default"])(_path$split),
          tail = _path$split2.slice(0);
        tail.reduce(function (acc, it) {
          if (!rootFound) {
            rootPaths.forEach(function (rootPath) {
              rootFound = rootFound || acc.startsWith(rootPath) && rootPath;
            });
          }
          if (rootFound && !openPaths.includes(acc)) {
            openPaths.push(acc);
            if (!openPaths.includes(rootFound)) {
              openPaths.push(rootFound);
            }
          }
          return acc + '/' + it;
        }, '');
      });
      return openPaths;
    }
  }, {
    key: "openPaths",
    value: function openPaths(paths) {
      var _this3 = this;
      if (!(paths instanceof Array)) {
        paths = [paths];
      }
      this.setState(function (prevState) {
        var openPaths = _this3.addPathToOpenPath(paths, _this3.props.rootPaths, prevState.openPaths);
        return {
          openPaths: openPaths
        };
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this4 = this;
      var selectedPaths = this.state.selectedPaths ? this.state.selectedPaths : this.props.selectedPaths;
      var openPaths = this.state.openPaths || this.props.openPaths;
      var setRefetch = this.props.setRefetch;
      openPaths = clone(openPaths);
      var vars = this.getVariables(selectedPaths, openPaths);
      return /*#__PURE__*/_react["default"].createElement(_reactComponents.Query, {
        query: this.query,
        variables: vars,
        fetchPolicy: "cache-first"
      }, function (result) {
        var error = result.error,
          loading = result.loading,
          data = result.data,
          refetch = result.refetch;
        if (setRefetch) {
          setRefetch({
            query: _this4.query,
            queryParams: vars,
            refetch: refetch
          });
        }
        var renderProp = _this4.props.children;
        if (_this4.props.onLoading) {
          _this4.props.onLoading(loading);
        }
        if (loading) {
          if (_this4.previousEntries) {
            return renderProp(_objectSpread({
              pickerEntries: _this4.previousEntries,
              loading: loading
            }, _this4.eventsHandlers));
          }
          return renderProp(_objectSpread({
            pickerEntries: [],
            loading: loading
          }, _this4.eventsHandlers));
        }
        if (error) {
          return renderProp(_objectSpread({
            pickerEntries: [],
            error: error,
            loading: loading
          }, _this4.eventsHandlers));
        }
        var pickerEntries = _this4.getPickerEntries(data, selectedPaths, openPaths);
        _this4.previousEntries = pickerEntries;
        return renderProp(_objectSpread({
          pickerEntries: pickerEntries,
          loading: loading
        }, _this4.eventsHandlers));
      });
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, prevState) {
      if (prevState.isOpenControlled !== Boolean(nextProps.openPaths) || prevState.isSelectControlled !== Boolean(nextProps.selectedPaths)) {
        console.warn('Cannot change between controlled/uncontrolled modes');
      }
      var newState = {};
      if (prevState.isOpenControlled && !(0, _fastDeepEqual["default"])(nextProps.openPaths, prevState.openPaths)) {
        newState.openPaths = nextProps.openPaths;
      }
      if (prevState.isSelectControlled && !(0, _fastDeepEqual["default"])(nextProps.selectedPaths, prevState.selectedPaths)) {
        newState.selectedPaths = nextProps.selectedPaths;
      }
      if (newState.openPaths || newState.selectedPaths) {
        return newState;
      }
      return null;
    }
  }]);
  return Picker;
}(_react["default"].Component);
exports.Picker = Picker;
(0, _defineProperty2["default"])(Picker, "propTypes", {
  /**
       * Optional set of fragments to add to the graphQL query. Can be a string that identify a predefinedFragment or a fragment definition
       */
  fragments: _propTypes["default"].array,
  /**
       * List of root paths for the picker
       */
  rootPaths: _propTypes["default"].arrayOf(_propTypes["default"].string),
  onOpenItem: _propTypes["default"].func,
  onSelectItem: _propTypes["default"].func,
  /**
       * List of open folders in controlled mode
       */
  openPaths: _propTypes["default"].arrayOf(_propTypes["default"].string),
  /**
       * List of selected path in controlled mode
       */
  selectedPaths: _propTypes["default"].arrayOf(_propTypes["default"].string),
  /**
       * Preselected items path (uncontrolled mode)
       */
  defaultSelectedPaths: _propTypes["default"].arrayOf(_propTypes["default"].string),
  /**
       * Callback when the selection has changed
       */
  onSelectionChange: _propTypes["default"].func,
  /**
       * List of folder paths that are open by default (uncontrolled mode)
       */
  defaultOpenPaths: _propTypes["default"].arrayOf(_propTypes["default"].string),
  /**
       * List of node types that can be "opened" (folders)
       */
  openableTypes: _propTypes["default"].arrayOf(_propTypes["default"].string),
  /**
       * List of node types that can be "selected" (items)
       */
  selectableTypes: _propTypes["default"].arrayOf(_propTypes["default"].string),
  onLoading: _propTypes["default"].func.isRequired,
  /**
       * Optional set of variable to pass to the graphQL query, in order to fulfill fragments needs
       */
  queryVariables: _propTypes["default"].objectOf(_propTypes["default"].any),
  hideRoot: _propTypes["default"].bool,
  /**
       * Optional function which receives refetch function of the Query component when the component is strapped
       */
  setRefetch: _propTypes["default"].func.isRequired,
  children: _propTypes["default"].func.isRequired
});
