// @jahia/moonstone/dist/components/Chip/Chip.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["label", "color", "icon", "isDisabled", "className"];
import React from 'react';
import clsx from 'clsx';
import "./Chip.css";
import { Typography } from "../Typography";
export var Chip = function Chip(_ref) {
  var _ref$label = _ref.label,
      label = _ref$label === void 0 ? '' : _ref$label,
      _ref$color = _ref.color,
      color = _ref$color === void 0 ? 'default' : _ref$color,
      _ref$icon = _ref.icon,
      icon = _ref$icon === void 0 ? null : _ref$icon,
      _ref$isDisabled = _ref.isDisabled,
      isDisabled = _ref$isDisabled === void 0 ? false : _ref$isDisabled,
      className = _ref.className,
      props = _objectWithoutProperties(_ref, _excluded);

  return /*#__PURE__*/React.createElement("div", _extends({
    className: clsx('moonstone-chip', "moonstone-color_".concat(color), {
      'moonstone-disabled': isDisabled
    }, className)
  }, props), icon && /*#__PURE__*/React.createElement(icon.type, {
    size: "small"
  }), label && /*#__PURE__*/React.createElement(Typography, {
    isNowrap: true,
    component: "span",
    variant: "caption",
    weight: "semiBold"
  }, label));
};
Chip.displayName = 'Chip';
