// @jahia/data-helper/dist/esm/hooks/useNodeInfo/useNodeInfo.gql-queries.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";
var _templateObject, _templateObject2, _templateObject3, _templateObject4;
import gql from 'graphql-tag';
import { aggregatedPublicationInfo, aggregatedPublicationInfoWithExistInLive, childNodeTypes, contentRestrictions, displayableNode, displayName, getProperties, installedModules, lockInfo, mimeTypes, nodeCacheRequiredFields, operationSupport, parentNode, primaryNodeType, replaceFragmentsInDocument, siteLanguages, subNodesCount } from "../../fragments";
import { getPermissionFragment, getSitePermissionFragment } from "../../fragments/getPermissionFragment";
import { getNodeTypeFragment } from "../../fragments/getIsNodeTypeFragment";
var getBaseQueryAndVariables = function getBaseQueryAndVariables(variables) {
  if (variables.paths) {
    return {
      baseQuery: gql(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n                query NodesByPathInfoQuery($paths:[String!]!) {\n                    jcr {\n                        nodesByPath(paths:$paths) {\n                            name\n                            ...node\n                            ...NodeCacheRequiredFields\n                        }\n                    }\n                }\n                ", "\n            "])), nodeCacheRequiredFields.gql),
      generatedVariables: {
        paths: variables.paths
      },
      skip: false
    };
  }
  if (variables.uuid && variables.uuid.length > 0) {
    return {
      baseQuery: gql(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n                query NodeByUuidInfoQuery($uuid:String!) {\n                    jcr {\n                        nodeById(uuid:$uuid) {\n                            name\n                            ...node\n                            ...NodeCacheRequiredFields\n                        }\n                    }\n                }\n                ", "\n            "])), nodeCacheRequiredFields.gql),
      generatedVariables: {
        uuid: variables.uuid
      },
      skip: false
    };
  }
  if (variables.uuids) {
    return {
      baseQuery: gql(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n                query NodesByUuidInfoQuery($uuids:[String!]!) {\n                    jcr {\n                        nodesById(uuids:$uuids) {\n                            name\n                            ...node\n                            ...NodeCacheRequiredFields\n                        }\n                    }\n                }\n                ", "\n            "])), nodeCacheRequiredFields.gql),
      generatedVariables: {
        uuids: variables.uuids
      },
      skip: false
    };
  }
  return {
    baseQuery: gql(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["\n            query NodeByPathInfoQuery($path:String!) {\n                jcr {\n                    nodeByPath(path:$path) {\n                        name\n                        ...node\n                        ...NodeCacheRequiredFields\n                    }\n                }\n            }\n            ", "\n        "])), nodeCacheRequiredFields.gql),
    generatedVariables: {
      path: variables.path
    },
    skip: !variables.path || variables.path.length === 0
  };
};
export var validOptions = ['getDisplayName', 'getPrimaryNodeType', 'getParent', 'getAggregatedPublicationInfo', 'getOperationSupport', 'getPermissions', 'getSitePermissions', 'getIsNodeTypes', 'getProperties', 'getSiteInstalledModules', 'getSiteLanguages', 'getDisplayableNodePath', 'getLockInfo', 'getChildNodeTypes', 'getContributeTypesRestrictions', 'getSubNodesCount', 'getMimeType'];
export var getQuery = function getQuery(variables, schemaResult) {
  var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
  var fragments = [];
  var _getBaseQueryAndVaria = getBaseQueryAndVariables(variables),
    baseQuery = _getBaseQueryAndVaria.baseQuery,
    generatedVariables = _getBaseQueryAndVaria.generatedVariables,
    skip = _getBaseQueryAndVaria.skip;
  if (!skip) {
    if (options.getDisplayName) {
      fragments.push(displayName);
      if (!variables.language) {
        throw Error('language is required');
      }
      generatedVariables.language = variables.language;
    }
    if (options.getPrimaryNodeType) {
      fragments.push(primaryNodeType);
      if (variables.displayLanguage) {
        generatedVariables.displayLanguageSet = true;
        generatedVariables.displayLanguage = variables.displayLanguage;
      } else {
        generatedVariables.displayLanguageSet = false;
        generatedVariables.displayLanguage = '';
      }
    }
    if (options.getParent) {
      fragments.push(parentNode);
    }
    if (options.getAggregatedPublicationInfo) {
      var supportsExistsInLive = schemaResult && schemaResult.__type && schemaResult.__type.fields && schemaResult.__type.fields.find(function (field) {
        return field.name === 'existsInLive';
      }) !== undefined;
      if (supportsExistsInLive) {
        fragments.push(aggregatedPublicationInfoWithExistInLive);
      } else {
        fragments.push(aggregatedPublicationInfo);
      }
      if (!variables.language) {
        throw Error('language is required');
      }
      generatedVariables.language = variables.language;
      generatedVariables.aggregatedPublicationInfoSubNodes = Boolean(options.getAggregatedPublicationInfo.subNodes);
      generatedVariables.aggregatedPublicationInfoIncludeReference = Boolean(options.getAggregatedPublicationInfo.reference);
    }
    if (options.getOperationSupport) {
      fragments.push(operationSupport);
    }
    if (options.getPermissions) {
      options.getPermissions.forEach(function (name) {
        var _getPermissionFragmen = getPermissionFragment(name),
          fragment = _getPermissionFragmen.fragment,
          fragmentVariables = _getPermissionFragmen.variables;
        fragments.push(fragment);
        Object.assign(generatedVariables, fragmentVariables);
      });
    }
    if (options.getSitePermissions) {
      options.getSitePermissions.forEach(function (name) {
        var _getSitePermissionFra = getSitePermissionFragment(name),
          fragment = _getSitePermissionFra.fragment,
          fragmentVariables = _getSitePermissionFra.variables;
        fragments.push(fragment);
        Object.assign(generatedVariables, fragmentVariables);
      });
    }
    if (options.getIsNodeTypes) {
      options.getIsNodeTypes.forEach(function (name) {
        var _getNodeTypeFragment = getNodeTypeFragment(name),
          fragment = _getNodeTypeFragment.fragment,
          fragmentVariables = _getNodeTypeFragment.variables;
        fragments.push(fragment);
        Object.assign(generatedVariables, fragmentVariables);
      });
    }
    if (options.getProperties) {
      fragments.push(getProperties);
      generatedVariables.getPropertiesNames = options.getProperties;
      if (!variables.language) {
        throw Error('language is required');
      }
      generatedVariables.language = variables.language;
    }
    if (options.getSiteInstalledModules) {
      fragments.push(installedModules);
    }
    if (options.getSiteLanguages) {
      fragments.push(siteLanguages);
    }
    if (options.getDisplayableNodePath) {
      fragments.push(displayableNode);
    }
    if (options.getLockInfo) {
      fragments.push(lockInfo);
    }
    if (options.getChildNodeTypes) {
      fragments.push(childNodeTypes);
    }
    if (options.getContributeTypesRestrictions) {
      fragments.push(contentRestrictions);
    }
    if (options.getSubNodesCount) {
      fragments.push(subNodesCount);
      generatedVariables.subNodesCountTypes = options.getSubNodesCount.types ? options.getSubNodesCount.types : ['nt:base'];
    }
    if (options.getMimeType) {
      fragments.push(mimeTypes);
    }
  }
  return {
    query: replaceFragmentsInDocument(baseQuery, fragments),
    generatedVariables: generatedVariables,
    skip: skip
  };
};
