// @jahia/data-helper/dist/esm/fragments/PredefinedFragments.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";
var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7, _templateObject8, _templateObject9, _templateObject10, _templateObject11, _templateObject12, _templateObject13, _templateObject14, _templateObject15, _templateObject16, _templateObject17, _templateObject18, _templateObject19, _templateObject20;
import gql from 'graphql-tag';
export var displayName = {
  variables: {
    language: 'String!'
  },
  applyFor: 'node',
  gql: gql(_templateObject || (_templateObject = _taggedTemplateLiteral(["fragment DisplayName on JCRNode {\n        displayName(language:$language)\n    }"])))
};
export var primaryNodeType = {
  variables: {
    displayLanguage: 'String!',
    displayLanguageSet: 'Boolean!'
  },
  applyFor: 'node',
  gql: gql(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["fragment NodeInfoPrimaryNodeType on JCRNode {\n        primaryNodeType {\n            name\n            displayName(language: $displayLanguage) @include(if: $displayLanguageSet)\n            icon\n        }\n    }"])))
};
export var parentNode = {
  applyFor: 'node',
  gql: gql(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["fragment ParentNodeInfo on JCRNode {\n        parent {\n            path\n            name\n            ...NodeCacheRequiredFields\n        }\n    }"])))
};
export var aggregatedPublicationInfo = {
  variables: {
    language: 'String!',
    aggregatedPublicationInfoSubNodes: 'Boolean',
    aggregatedPublicationInfoIncludeReference: 'Boolean'
  },
  applyFor: 'node',
  gql: gql(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["fragment AggregatedPublicationInfo on JCRNode {\n        aggregatedPublicationInfo(language: $language, subNodes: $aggregatedPublicationInfoSubNodes, references:$aggregatedPublicationInfoIncludeReference) {\n            publicationStatus\n        }\n    }"])))
};
export var aggregatedPublicationInfoWithExistInLive = {
  variables: {
    language: 'String!',
    aggregatedPublicationInfoSubNodes: 'Boolean',
    aggregatedPublicationInfoIncludeReference: 'Boolean'
  },
  applyFor: 'node',
  gql: gql(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["fragment AggregatedPublicationInfoWithExistsInLive on JCRNode {\n        aggregatedPublicationInfo(language: $language, subNodes: $aggregatedPublicationInfoSubNodes, references:$aggregatedPublicationInfoIncludeReference) {\n            publicationStatus\n            existsInLive\n        }\n    }"])))
};
export var operationSupport = {
  applyFor: 'node',
  gql: gql(_templateObject6 || (_templateObject6 = _taggedTemplateLiteral(["fragment OperationSupport on JCRNode {\n        operationsSupport {\n            lock\n            markForDeletion\n            publication\n        }\n    }"])))
};
export var allProperties = {
  variables: {
    language: 'String!'
  },
  applyFor: 'node',
  gql: gql(_templateObject7 || (_templateObject7 = _taggedTemplateLiteral(["fragment NodeAllProperties on JCRNode {\n        properties(language:$language) {\n            name\n            value\n            values\n        }\n    }"])))
};
export var getProperties = {
  variables: {
    language: 'String!',
    getPropertiesNames: '[String!]!'
  },
  applyFor: 'node',
  gql: gql(_templateObject8 || (_templateObject8 = _taggedTemplateLiteral(["fragment NodeProperties on JCRNode {\n        properties(names: $getPropertiesNames, language: $language) {\n            name\n            value\n            values\n        }\n    }"])))
};
export var installedModules = {
  applyFor: 'node',
  gql: gql(_templateObject9 || (_templateObject9 = _taggedTemplateLiteral(["fragment SiteInstalledModules on JCRNode {\n        site {\n            installedModulesWithAllDependencies\n            ...NodeCacheRequiredFields\n        }\n    }"])))
};
export var siteLanguages = {
  applyFor: 'node',
  gql: gql(_templateObject10 || (_templateObject10 = _taggedTemplateLiteral(["fragment SiteLanguages on JCRNode {\n        site {\n            defaultLanguage\n            ...NodeCacheRequiredFields\n            languages {\n                displayName\n                language\n                activeInEdit\n            }\n        }\n    }"])))
};
export var displayableNode = {
  applyFor: 'node',
  gql: gql(_templateObject11 || (_templateObject11 = _taggedTemplateLiteral(["fragment DisplayableNodePath on JCRNode {\n        displayableNode {\n            path\n            ...NodeCacheRequiredFields\n        }\n    }"])))
};
export var lockInfo = {
  applyFor: 'node',
  gql: gql(_templateObject12 || (_templateObject12 = _taggedTemplateLiteral(["fragment LockInfo on JCRNode {\n        lockOwner: property(name: \"jcr:lockOwner\") {\n            value\n        }\n        lockTypes: property(name: \"j:lockTypes\") {\n            values\n        }\n    }"])))
};
export var subNodesCount = {
  variables: {
    subNodesCountTypes: '[String!]!'
  },
  applyFor: 'node',
  gql: gql(_templateObject13 || (_templateObject13 = _taggedTemplateLiteral(["fragment SubNodesCount on JCRNode {\n        subNodes: children(typesFilter: {types: $subNodesCountTypes, multi: ANY}) {\n            pageInfo {\n                totalCount\n            }\n        }\n    }"])))
};
export var childNodeTypes = {
  applyFor: 'node',
  gql: gql(_templateObject14 || (_templateObject14 = _taggedTemplateLiteral(["fragment AllowedChildNodeType on JCRNode {\n        allowedChildNodeTypes(includeSubTypes: false) {\n            name\n        }\n    }"])))
};
export var contentRestrictions = {
  applyFor: 'node',
  gql: gql(_templateObject15 || (_templateObject15 = _taggedTemplateLiteral(["fragment ContentRestriction on JCRNode {\n        contributeTypes: property(name: \"j:contributeTypes\") {\n            values\n        }\n        ancestors(fieldFilter: {filters: {evaluation: NOT_EMPTY, fieldName: \"contributeTypes\"}}) {\n            ...NodeCacheRequiredFields\n            contributeTypes : property(name: \"j:contributeTypes\") {\n                values\n            }\n        }\n    }"])))
};
export var siteHomePage = {
  applyFor: 'node',
  gql: gql(_templateObject16 || (_templateObject16 = _taggedTemplateLiteral(["fragment NodeSiteHomePage on JCRNode {\n        children(typesFilter:{types:[\"jnt:page\"]}, propertiesFilter:{filters:[{property:\"j:isHomePage\", value:\"true\" }]}) {\n            nodes {\n                path\n                ...NodeCacheRequiredFields\n            }\n        }\n    }"])))
};
export var nodeCacheRequiredFields = {
  applyFor: 'node',
  gql: gql(_templateObject17 || (_templateObject17 = _taggedTemplateLiteral(["fragment NodeCacheRequiredFields on JCRNode {\n        uuid\n        workspace\n        path\n    }"])))
};
export var nodeTypeDisplayName = {
  variables: {
    language: 'String!'
  },
  applyFor: 'nodeType',
  gql: gql(_templateObject18 || (_templateObject18 = _taggedTemplateLiteral(["fragment NodeTypeDisplayName on JCRNodeType {\n        displayName(language:$language)\n    }"])))
};
export var nodeTypeSubTypes = {
  variables: {
    language: 'String!'
  },
  applyFor: 'nodeType',
  gql: gql(_templateObject19 || (_templateObject19 = _taggedTemplateLiteral(["fragment NodeTypeSubTypes on JCRNodeType {\n        subTypes {\n            nodes {\n                name\n                displayName(language:$language)\n                abstract\n                mixin\n            }\n        }\n    }"])))
};
export var mimeTypes = {
  applyFor: 'node',
  gql: gql(_templateObject20 || (_templateObject20 = _taggedTemplateLiteral(["fragment NodeInfoResourceNode on JCRNode {\n        resourceChildren: children(names: [\"jcr:content\"]) {\n            nodes {\n                ...NodeCacheRequiredFields\n                mimeType: property(name: \"jcr:mimeType\") {\n                    value\n                }\n            }\n        }\n    }"])))
};
export var PredefinedFragments = {
  displayName: displayName,
  primaryNodeType: primaryNodeType,
  parentNode: parentNode,
  aggregatedPublicationInfo: aggregatedPublicationInfo,
  operationSupport: operationSupport,
  allProperties: allProperties,
  getProperties: getProperties,
  installedModules: installedModules,
  siteLanguages: siteLanguages,
  displayableNode: displayableNode,
  lockInfo: lockInfo,
  subNodesCount: subNodesCount,
  contentRestrictions: contentRestrictions,
  siteHomePage: siteHomePage,
  nodeCacheRequiredFields: nodeCacheRequiredFields,
  nodeTypeDisplayName: nodeTypeDisplayName,
  nodeTypeSubTypes: nodeTypeSubTypes,
  mimeTypes: mimeTypes
};
