// @jahia/moonstone/dist/components/Badge/Badge.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["label", "color", "className"];
import React from 'react';
import "./Badge.css";
import { Typography } from "../Typography";
import clsx from 'clsx';
export var Badge = function Badge(_ref) {
  var _ref$label = _ref.label,
      label = _ref$label === void 0 ? null : _ref$label,
      _ref$color = _ref.color,
      color = _ref$color === void 0 ? 'accent' : _ref$color,
      className = _ref.className,
      other = _objectWithoutProperties(_ref, _excluded);

  var classNameProps = clsx('moonstone-badge', 'moonstone-badge_round', "moonstone-badge_".concat(color), className);

  if (!label || label.length < 1) {
    return null;
  }

  return /*#__PURE__*/React.createElement(Typography, _extends({
    isNowrap: true,
    component: "span",
    variant: "caption",
    weight: "bold",
    className: classNameProps
  }, other), label);
};
Badge.displayName = 'Badge';
