// @jahia/moonstone/dist/components/Input/SearchContextInput/SearchContextInput.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["searchContext"];
import React from 'react';
import { BaseInput } from "../BaseInput";
import clsx from 'clsx';
export var SearchContextInput = function SearchContextInput(_ref) {
  var searchContext = _ref.searchContext,
      props = _objectWithoutProperties(_ref, _excluded);

  var hasSearchContext = typeof searchContext !== 'undefined';
  var component = hasSearchContext && /*#__PURE__*/React.createElement(searchContext.type, _extends({
    key: "searchcontext"
  }, searchContext.props, {
    variant: "ghost",
    size: "small",
    className: clsx(searchContext.props.className, 'moonstone-searchContextInput_element')
  }));
  return /*#__PURE__*/React.createElement(BaseInput, _extends({}, props, {
    isShowClearButton: true,
    prefixComponents: [component],
    icon: null,
    role: "search",
    size: "big"
  }));
};
SearchContextInput.displayName = 'SearchContextInput';
